### PR TITLE
Show warning for pipelines configured with 'set_pipeline' step

### PR DESCRIPTION
**Pull request by Infra-Red**
[imperative mood]: https://chris.beams.io/posts/git-commit/#imperative

## What does this PR accomplish?

Bug Fix | **Feature** | Documentation

closes https://github.com/concourse/concourse/issues/5814.

## Changes proposed by this PR:

This PR resolves the last missing piece to complete [RFC #31](https://github.com/concourse/rfcs/blob/master/031-set-pipeline-step/proposal.md). With this change `fly set-pipeline` will notify the operator that the pipeline has been configured with `set_pipeline` step and all changes may be lost on the next `set_pipeline` step execution. 

<details><summary>set-pipeline command output</summary>
<p>

```
❯ ./fly-darwin -t local sp -p hello -c pipeline.yml
resource types:
  resource type registry-image-resource has changed:
  name: registry-image-resource
  source:
-   repository: xxx/registry-image-resource1
+   repository: xxx/registry-image-resource
    tag: latest
  type: registry-image

pipeline name: hello

WARNING: pipeline has a 'set_pipeline' step configured, your changes may be blown away on the next 'step_pipeline' step execution

apply configuration? [yN]: y
configuration updated
```

</p>
</details>

## Notes to reviewer:

Let me know if all 'experimental' annotations related to `set_pipeline` step can be removed as a part of this PR.

## Release Note

* `fly set-pipeline` now prints warning message when a pipeline has been configured through the `set_pipeline` step.

## Contributor Checklist

- [X] Followed [Code of conduct], [Contributing Guide] & avoided [Anti-patterns]
- [X] [Signed] all commits
- [X] Added tests (Unit and/or Integration)
- [ ] Updated [Documentation]
- [ ] Added release note (Optional)

[Code of Conduct]: https://github.com/concourse/concourse/blob/master/CODE_OF_CONDUCT.md
[Contributing Guide]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md
[Anti-patterns]: https://github.com/concourse/concourse/wiki/Anti-Patterns
[Signed]: https://help.github.com/en/github/authenticating-to-github/signing-commits
[Documentation]: https://github.com/concourse/docs

## Reviewer Checklist

- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the
  [BOSH](https://github.com/concourse/concourse-bosh-release) and
  [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for
  the [integration
  tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env)
  (for example, if they are Garden configs that are not displayed in the
  `--help` text).

### CODE REVIEW

**Comment by aoldershaw** [comment]
Thanks for the PR! I think there may have been a slight misunderstanding around the manual update feature, though - sorry about that! [Preventing manual updates](https://github.com/concourse/rfcs/blob/master/031-set-pipeline-step/proposal.md#preventing-manual-updates) states:

> When using `fly set-pipeline` to update a pipeline that has been configured through the `set_pipeline` step, a warning will be printed and a confirmation dialogue will be presented

So, if you run

```yaml
$ fly -t dev set-pipeline -p my-pipeline -c config.yml
```

we don't care whether `config.yml` has a `set_pipeline` step in it - we only print a warning if `my-pipeline` was previously set by another pipeline's `set_pipeline` step (or I suppose `my-pipeline` set itself via `set_pipeline: self` after being manually set prior).

For pipelines set by a `set_pipeline` step, we currently track the `build_id` that most recently called `set_pipeline`, and the `job_id` that build belongs to, whereas manually set pipelines do not contain that information - see
https://github.com/concourse/concourse/blob/4c5121cb96c4299ee4330114ec70fdc3dfb7e102/atc/exec/set_pipeline_step.go#L218
and
https://github.com/concourse/concourse/blob/4c5121cb96c4299ee4330114ec70fdc3dfb7e102/atc/db/team.go#L429-L430

So, we're able to determine if a pipeline was set via the `set_pipeline` step, but we don't expose this information anywhere in the API. I suspect the best place for this would be in the get pipeline endpoint: see
https://github.com/concourse/concourse/blob/4c5121cb96c4299ee4330114ec70fdc3dfb7e102/atc/api/pipelineserver/get.go#L17
and
https://github.com/concourse/concourse/blob/4c5121cb96c4299ee4330114ec70fdc3dfb7e102/atc/pipeline.go#L14-L25

We could add `ParentBuildID` and `ParentJobID` to `atc.Pipeline`. We'd also have to make another API call to the `Pipeline` endpoint prior to updating the pipeline config, like we do here: https://github.com/concourse/concourse/blob/4c5121cb96c4299ee4330114ec70fdc3dfb7e102/fly/commands/internal/setpipelinehelpers/atc_config.go#L110

...and check whether `ParentBuildID` and `ParentJobID` are set - if so, add a warning.

Let me know if that makes sense!

**Comment by Infra-Red** [comment]
Hi @aoldershaw! Thanks for clarifications and apologize that I misunderstood the requirement.

**Comment by Infra-Red** [comment]
Hi @aoldershaw! Thanks for clarifications and apologize that I misunderstood the requirement. 

I want to clarify one thing. Is it correct that `ParentBuildID` and `ParentJobID` fields are only set if pipeline was configured using `set_pipeline` step?

Thanks again for a detailed explanation. I will make the necessary changes in this PR.

**Comment by aoldershaw** [comment]
> I want to clarify one thing. Is it correct that `ParentBuildID` and `ParentJobID` fields are only set if pipeline was configured using `set_pipeline` step?

Yes, that's correct!

**Comment by Infra-Red** [comment]
Hi @aoldershaw! I updated this PR according to your suggestions. Please review when you get a chance.

**Review by aoldershaw — CHANGES_REQUESTED** [review]
Looks great, works as I'd expect - pipelines without a parent don't print a warning, pipelines with a parent print a warning, and if you manually set a pipeline that once had a parent, you don't receive the warning if you try to set it again. Thanks for making the change!

Just one small typo, but other than that, this is good to go!

File: fly/integration/set_pipeline_test.go [hunk 1]
```diff
@@ -1245,6 +1258,52 @@ this is super secure
...
+					Expect(func() {
+						flyCmd := exec.Command(flyPath, "-t", targetName, "set-pipeline", "-p", "awesome-pipeline", "-c", configFile.Name())
+
+						stdin, err := flyCmd.StdinPipe()
+						Expect(err).NotTo(HaveOccurred())
+
+						sess, err := gexec.Start(flyCmd, GinkgoWriter, GinkgoWriter)
+						Expect(err).NotTo(HaveOccurred())
+
+						Eventually(sess).Should(gbytes.Say("WARNING: pipeline has been configured through the 'set_pipeline' step, your changes may be overwritten on the next 'step_pipeline' step execution"))
```
Comment (aoldershaw) [inline fly/integration/set_pipeline_test.go:1294]: @Infra-Red thanks for the quick fix - but looks like the test needs updating too

Comment (Infra-Red) [reply fly/integration/set_pipeline_test.go]: The test was also fixed now.

**Review by aoldershaw — APPROVED** [review]
Thanks!

**Pull request merged**